### PR TITLE
tsdbrelay: denormalization backfill app

### PR DESCRIPTION
**Pull request by captncraig**
[Review on Reviewable](https://reviewable.io/reviews/bosun-monitor/bosun/1076)

### CODE REVIEW

**Comment by mjibson** [comment]
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 1 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1RD4nihRxPdEW1s4K)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L1)):</sup>
Why isn't this program in the cmd directory with all the other commands?

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 15 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1Svz0_kEwt2hL_PrS)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L15)):</sup>
Why do we need q at all? Can't we infer this from ruleFlag completely?

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 16 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1SevBNNU1Zqy3QIlb)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L16)):</sup>
I'd recommend YYYY-MM-DD here instead of slashes. Dashes are generally used if year is first, slashes if year is last. Maybe because that's OpenTSDB's date format?

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 59 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1sQNeBzOJ4KB8W0IA)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L59)):</sup>
I think this can be omitted because it doesn't tell the user anything they didn't know.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 63 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1uAcds7D7RbdCtUGW)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L63)):</sup>
What happens if the queue length is too small?

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 68 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1sa7mYRAmv92xRfTr)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L68)):</sup>
This URL should be specified on the command line.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 109 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1u6NZQrxuABpMNw0o)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L109)):</sup>
Why this instead of sending it ourselves? Because of the retry logic in collect, or some other reason?

tsdbrelay just ignores failed denormalize requests: https://github.com/bosun-monitor/bosun/blob/1cc2fd61d4046a7dbd663d436741b2f44db5226e/cmd/tsdbrelay/main.go#L184

Maybe it should have some retry logic too?

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 128 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1wA7FW_NkrDRapr7A)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L128)):</sup>
This could cause the backfill process to exit before all the datapoints have been sent. See: https://github.com/bosun-monitor/bosun/blob/1cc2fd61d4046a7dbd663d436741b2f44db5226e/collect/queue.go#L49

Points are removed from the queue and then the send attempt is made. If it fails, then they are requeued. If we are only checking to see if the queue length is 0, then we could exit while there are points in-flight. This could cause the request to not be sent (since the entire program exits), or a failed batch to not be requeued.

---

<sup>**[collect/collect.go, line 116 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1voh2tKMa0-w1C95G)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/collect/collect.go#L116)):</sup>
I do not like that we start 3 goroutines and then do a check that can return an error. Do all possible things that can return an error before starting the goroutines.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by captncraig** [comment]
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 1 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1RD4nihRxPdEW1s4K)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L1)):</sup>
just for naming I liked it being under the denormalize namespace. Also didn't want to overpopulate the cmd namespace. But I think you are right.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 17 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1Svz0_kEwt2hL_PrS)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L17)):</sup>
the rule might be `os.net.bytes__host` but we would still want to query `os.net.bytes{host=*,direction=*,interface=*}` to get all points. I was thinking of auto-divining all of the aggregate tags, but it seemed easier in the end to just make the query explicit.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 18 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1SevBNNU1Zqy3QIlb)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L18)):</sup>
This is the opentsdb format.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 63 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1uAcds7D7RbdCtUGW)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L63)):</sup>
If the queue overflows, we will drop stuff. That would not be good. Perhaps I should check the queue between time intervals to be safe.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 68 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1sa7mYRAmv92xRfTr)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L68)):</sup>
Oops.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 109 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1u6NZQrxuABpMNw0o)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L109)):</sup>
It was an attempt to not duplicate the gzip, json code once again. Maybe I should just centralize the "send a bunch of data points to opentsdb" bit in the collect package.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 128 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1wA7FW_NkrDRapr7A)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/cmd/tsdbrelay/denormalize/backfill/main.go#L128)):</sup>
using collect was an attempt to save code, but I agree I probably don't want all of the baggage. I'll make the relevant bits reusable instead of using the entire collect pipeline.

---

<sup>**[collect/collect.go, line 116 \[r3\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1voh2tKMa0-w1C95G)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d8447f6369cc94488ab99827830f73e8f37dd75d/collect/collect.go#L116)):</sup>
true. I'll revert this. The reasoning was if you are not sending self metrics, the root doesn't matter.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by mjibson** [comment]
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/tsdbrelay/denormalize/backfill/main.go, line 17 \[r1\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1Svz0_kEwt2hL_PrS)** ([raw file](https://github.com/bosun-monitor/bosun/blob/ebecaf6f5166a23ebc74b6d4fa2e1a553e8e7d14/cmd/tsdbrelay/denormalize/backfill/main.go#L17)):</sup>
Making the query explicit allows a user to accidentally do something wrong (i.e., they didn't know about the interface tag and now their backfill is useless since all their incoming data has it). OpenTSDB reports back the aggregate tags, so we can use that to make the correct query. I think it's worth the work.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

File: collect/queue.go [hunk 1]
```diff
@@ -136,3 +111,26 @@ func recordSent(num int) {
...
+	req, err := http.NewRequest("POST", tsdb, &buf)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Content-Encoding", "gzip")
+
+	resp, err := client.Do(req)
+	if err == nil {
+		defer resp.Body.Close()
```
Comment (mjibson) [inline collect/queue.go]: This breaks the error handling in `send()`, which reads from the body in case the status code isn't 204.

**Comment by mjibson** [comment]
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/backfill/main.go, line 1 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js69c0TpGPuLhOzJoZZ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L1)):</sup>
Package comments should start with the name of the program: "Backfill denormalizes historic OpenTSDB data.".

---

<sup>**[cmd/backfill/main.go, line 2 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js69xkDf4-ShPmJol8i)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L2)):</sup>
This should have a newline above it to prevent this sentence from appearing in the godoc index.

---

<sup>**[cmd/backfill/main.go, line 21 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CNMVWyh5FQO9YHsZ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L21)):</sup>
opentsdb -> OpenTSDB

---

<sup>**[cmd/backfill/main.go, line 22 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CSuckePkeQWKlHMj)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L22)):</sup>
OpenTSDB

---

<sup>**[cmd/backfill/main.go, line 52 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CkJ6wK7RbB-UeKGK)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L52)):</sup>
can remove this debug

---

<sup>**[cmd/backfill/main.go, line 112 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6D5jpgg1tVMID2hh3)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L112)):</sup>
Isn't 204 expected, not 200?

---

<sup>**[cmd/backfill/main.go, line 139 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DYsicdGgFqXUgMap)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L139)):</sup>
return err

---

<sup>**[cmd/backfill/main.go, line 142 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DWWq_rkTZN8yXZ1_)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L142)):</sup>
return err

---

<sup>**[collect/collect.go, line 107 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DwYpwVpoQTQMXjen)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/collect/collect.go#L107)):</sup>
Why did this move?

---

<sup>**[collect/collect.go, line 109 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6Dv08fXWMIABGVl57)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/collect/collect.go#L109)):</sup>
I don't understand why this block moved.

---

<sup>**[collect/collect.go, line 120 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DyQXKMWV2351I8Qw)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/collect/collect.go#L120)):</sup>
Why the extra newline?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by captncraig** [comment]
Review status: 0 of 9 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/backfill/main.go, line 1 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js69c0TpGPuLhOzJoZZ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L1)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 2 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js69xkDf4-ShPmJol8i)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L2)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 21 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CNMVWyh5FQO9YHsZ)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L21)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 22 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CSuckePkeQWKlHMj)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L22)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 52 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6CkJ6wK7RbB-UeKGK)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L52)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 112 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6D5jpgg1tVMID2hh3)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L112)):</sup>
from http://opentsdb.net/docs/build/html/api_http/query/index.html it looks like 200 is expected. That is what I get from devbosun anyway. Should I look for 2xx instead?

---

<sup>**[cmd/backfill/main.go, line 139 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DYsicdGgFqXUgMap)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L139)):</sup>
Done.

---

<sup>**[cmd/backfill/main.go, line 142 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DWWq_rkTZN8yXZ1_)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L142)):</sup>
Done.

---

<sup>**[collect/collect.go, line 107 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6DwYpwVpoQTQMXjen)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/collect/collect.go#L107)):</sup>
replaced

---

<sup>**[collect/queue.go, line 133 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js1UGY0P0QjuJ3fUvYz-r4-133)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/collect/queue.go#L133)):</sup>
Good point.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by mjibson** [comment]
Reviewed 5 of 6 files at r1, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[cmd/backfill/main.go, line 113 \[r5\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6b_UHlWckeb6PhIHP)** ([raw file](https://github.com/bosun-monitor/bosun/blob/014e2c4a8af95e6f308ba4fc964a0cbc6c6729e7/cmd/backfill/main.go#L113)):</sup>
no need for the else here since you have a return

---

<sup>**[cmd/backfill/main.go, line 118 \[r4\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-Js6D5jpgg1tVMID2hh3)** ([raw file](https://github.com/bosun-monitor/bosun/blob/dab78db472d1f6d68b5fd1b22d7e7e106d0c85bd/cmd/backfill/main.go#L118)):</sup>
Sorry, you're right. 200 is correct for queries, 204 is for /api/put.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by mjibson** [comment]
Reviewed 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[collect/queue.go, line 117 \[r7\]](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076#-JsBuHUx9pQipsfsW5jL)** ([raw file](https://github.com/bosun-monitor/bosun/blob/d5c62a8b4b1b014eeaabe9e8b7f58fe586468d8c/collect/queue.go#L117)):</sup>
I think the current changes will work. However, if it were me I'd just copy this function into the backfill program so we didn't have to change this file at all. Here's the reason: this function is only about 20 lines, and it's pretty simple. Yet it required about *40* other lines to change in order for the refactor to work. Just not worth it to me. Your choice.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by mjibson** [comment]
LGTM with the caveat I posted below.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bosun-monitor/bosun/1076)

**Comment by captncraig** [comment]
@mjibson I'm gonna merge as is. Most of the changes to collect were to change queue to carry datapoints instead of raw messages. This seems cosmetic now, but will be necessary in the future if scollector is to target more backends.

**Comment by mjibson** [comment]
Agreed. Will be useful for multiply back ends.

**Pull request merged**